### PR TITLE
Remove `eval` usage. Fixes CSP cases.

### LINKED ISSUE #335: 🐛 CSP 'unsafe-eval' not allowed
In my packaged non-minified build, there are eval calls, which yield the following EvalError in my specific use case, where 'unsafe-eval' Content Security Policy is not allowed.

```
Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'
```

**Is there any way I can fix unsafe eval calls, e.g. using a custom packager?**

My deps:

```
  "dependencies": {
    "babel-polyfill": "^6.26.0",
    "mo-js": "^0.288.2",
    "react": "^16.2.0",
    "react-dom": "^16.2.0",
    "uiw": "^1.5.3",
    "uiw-iconfont": "^1.2.6"
  },
  "devDependencies": {
    "babel-plugin-transform-class-properties": "^6.24.1",
    "babel-plugin-transform-object-rest-spread": "^6.26.0",
    "babel-preset-env": "^1.6.1",
    "babel-preset-react": "^6.24.1",
    "concurrently": "^3.5.1",
    "less": "^2.7.3",
    "node-sass": "^4.7.2",
    "parcel-bundler": "^1.2.0",
  }
```

.babelrc
```
{
  "presets": ["env", "react"],
  "plugins": ["transform-class-properties", "transform-object-rest-spread"]
}
```

For further context:

The application runs in an iframe with the following sandbox values:

```
sandbox="allow-forms allow-scripts allow-same-origin allow-popups allow-popups-to-escape-sandbox"
```

Thank you!

### PR DESCRIPTION
**Pull request by CvX**
Picks up on work done by @mBeierl (closes #353).
Fixes #335.

### CODE REVIEW

File: src/visitors/globals.js [hunk 1]
```diff
@@ -7,7 +7,8 @@ const VARS = {
     asset.addDependency('process');
     return 'var process = require("process");';
   },
-  global: () => 'var global = (1,eval)("this");',
+  global: () =>
+    'var global = typeof global === "object" ? global : typeof window === "object" ? window : typeof self === "object" ? self : {};',
```
Comment (devongovett) [inline src/visitors/globals.js]: I don't think this is necessarily very reliable. What if someone declares `window` or `self` in the local scope of their module?

Comment (CvX) [reply src/visitors/globals.js]: That's a good point. Replaced it with a more reliable (and succinct!) solution: https://github.com/parcel-bundler/parcel/pull/1133/commits/1fe753ed5d276eeaf8619100d6bd22aa15eb27fe

Comment (devongovett) [reply src/visitors/globals.js]: Yeah but won't `Function` constructor have the same problem as `eval` with CSP?

Comment (CvX) [reply src/visitors/globals.js]: You're, right, my bad.
I don't know why I assumed that another variant of arbitrary code execution would fly. 🙃  Should have read [W3C Working Draft for CSP](https://www.w3.org/TR/CSP/#directive-script-src)…

About the case of `window` and `self` in a local scope – you're talking about a situation where there's a module that either:
1. uses `global` object, redefines `window`, and is used in a browser
2. uses `global` object, redefines `self`, and is used in a Web Worker

Do we have to choose between supporting those two cases and supporting CSP?

Comment (devongovett) [reply src/visitors/globals.js]: Yeah... we may be able to get it working by moving that code to get the global object you originally added outside of each module and into the prelude where we can guarantee that nothing will mess with those variables. In there, the global object should be `this` (unless strict mode is enabled). Then we can attach it to `module` or something, and here assign `var global = module.global`, or maybe pass it into the module function. What do you think?

See this article for more info: https://www.contentful.com/blog/2017/01/17/the-global-object-in-javascript/

**Comment by CvX** [comment]
@devongovett Moving acquisition of `global` to the prelude is a great idea. I've updated the PR – let me know what you think. It's using `this` since prelude seems to be in a non-strict scope. Is there any case where it would become strict?

Btw, it would be a good idea to run some of the tests in an actual browser environment, maybe via Puppeteer. I'll have to look into that. Some actual CSP tests would be nice.

**Comment by devongovett** [comment]
There was one additional case to handle: when you redeclare `global` as a variable inside a module, e.g. `const global = 2`. This would cause a syntax error complaining that `global` was already defined (in the module function wrapper). I solved this by accessing `global` from `arguments[3]` when needed instead of always passing to the function.

**Comment by RichardJECooke** [comment]
I use latest Parcel 2 and I still see `eval` in my JS build. How do I disable the `eval` so I can use Parcel to build my browser script that runs in Electron without warnings please?

Package:

```js
  "devDependencies": {
    "electron": "^39.2.2",
    "parcel": "^2.16.1",
    "@parcel/transformer-vue": "^2.16.1"
  },
```

Output JS:

```js
// eval may also be disabled via CSP, so do a quick check.
    var supportsSourceURL = false;
    try {
        (0, eval)('throw new Error("test"); //# sourceURL=test.js');
    } catch (err) {
        supportsSourceURL = err.stack.includes('test.js');
    }
```

**Pull request merged**